### PR TITLE
Implement Kubernetes Selectors, minor kube endpoint fix

### PR DESCRIPTION
**Pull request by pnegahdar**

### CODE REVIEW

**Comment by errm** [comment]
Wow awesome thanks..

Might we be able to validate the label selector?

Oh and the build is failing because you need to run gofmt...

**Comment by pnegahdar** [comment]
Yeah as I mentioned in the issue the selectors have a special syntax that would be difficult to parse without importing the kube code. 

http://kubernetes.io/docs/user-guide/labels/

**Comment by emilevauge** [comment]
I tend to agree with @pnegahdar to NOT import k8s code :smile: 
Great PR btw!
Tests are still failling because of `go lint`: it would be better to test locally before using `make binary test-integration`.

File: provider/k8s/client.go [hunk 1]
```diff
@@ -227,10 +239,15 @@ func (c *clientImpl) watch(url string, stopCh <-chan bool) (chan interface{}, ch
...
-	request := c.request(url)
+	queryParams := map[string]string{"watch": "", "resourceVersion": resourceVersion}
+	if labelSelector != "" {
+		queryParams["labelSelector"] = labelSelector
+	}
+	queryData, err := json.Marshal(queryParams)
+	if err != nil {
+		return watchCh, errCh, fmt.Errorf("Unable to construct query args")
+	}
+	request := c.request(url, string(queryData))
```
Comment (errm) [inline provider/k8s/client.go]: This looks very similar to L57-65 could we extract a function to dry it up a little.

File: provider/kubernetes.go [hunk 2]
```diff
@@ -74,9 +76,15 @@ func (provider *Kubernetes) createClient() (k8s.Client, error) {
 	}
 	kubernetesHost := os.Getenv("KUBERNETES_SERVICE_HOST")
 	kubernetesPort := os.Getenv("KUBERNETES_SERVICE_PORT_HTTPS")
-	if len(kubernetesPort) > 0 && len(kubernetesHost) > 0 {
+	// Prioritize user provided kubernetes endpoint since kube container runtime will almost always have it
+	if provider.Endpoint == "" && len(kubernetesPort) > 0 && len(kubernetesHost) > 0 {
+		log.Debugf("Using environment provided kubernetes endpoint")
```
Comment (errm) [inline provider/kubernetes.go]: we could log the value of provider.Endpoint here...

**Comment by vdemeester** [comment]
@pnegahdar could you squash your commit too 👼

**Comment by pnegahdar** [comment]
Sure. 2 (one per issue) or just 1?

**Comment by emilevauge** [comment]
One per issue is perfect :)

**Comment by errm** [comment]
LGTM

File: configuration.go [hunk 3]
```diff
@@ -272,7 +272,8 @@ func NewTraefikDefaultPointersConfiguration() *TraefikConfiguration {
 	//default Kubernetes
 	var defaultKubernetes provider.Kubernetes
 	defaultKubernetes.Watch = true
-	defaultKubernetes.Endpoint = "http://127.0.0.1:8080"
+	defaultKubernetes.Endpoint = ""
```
Comment (emilevauge) [inline configuration.go]: Why did you remove this default value ?

Comment (pnegahdar) [inline configuration.go]: To know if it was explicitly provided to decide whether to use the
kubernetes provide environment variables. It falls back to the 127 default
otherwise

Comment (emilevauge) [inline configuration.go]: OK thanks

**Comment by emilevauge** [comment]
LGTM
